cwlVersion: v1.2
class: CommandLineTool
id: kallisto
requirements:
  - class: ShellCommandRequirement
  - class: DockerRequirement
    dockerPull: 'images.sbgenomics.com/uros_sipetic/kallisto:0.43.1'
  - class: InlineJavascriptRequirement
  - class: ResourceRequirement
    coresMin: 8
    ramMin: 10000

baseCommand: []
arguments:
  - position: 1
    shellQuote: false
    valueFrom: >-
      kallisto quant
      -i $(inputs.transcript_idx.path)
      -o output
      -b 10
      -t 8
      $(inputs.strand ? inputs.strand == "default" ? "" : "--"+inputs.strand : "")
      $(inputs.reads_records.every(function(e) { return !e.is_paired_end }) ? ["--single", "-l", inputs.avg_frag_len, "-s", inputs.std_dev].join(' ') : "")
      $(inputs.reads_records.map(function(e) { return (e.reads2 != null ? [e.reads1.path, e.reads2.path].join(' ') : e.reads1.path) }).join(' '))
      && mv output/abundance.tsv $(inputs.SampleID).kallisto.abundance.tsv
      && gzip $(inputs.SampleID).kallisto.abundance.tsv

inputs:
  transcript_idx: File
  strand: {type: ['null', string], doc: "input none if unstranded, otherwise rf-stranded or fr-stranded"}
  reads_records:
    type:
      type: array
      items:
        type: record
        fields:
          reads1: { type: File, doc: "Input fastq file, gzipped or uncompressed OR alignment file"}
          reads2: { type: 'File?', doc: "If paired end, R2 reads files, gzipped or uncompressed" }
          is_paired_end: { type: 'boolean?', doc: "Are the reads paired end?" }
          outSAMattrRGline: { type: 'string?', doc: "Suggested setting format is: ID:sample_name LB:aliquot_id PL:platform SM:BSID for example ID:7316-242 LB:750189 PL:ILLUMINA SM:BS_W72364M N. STAR will automatically convert unquoted spaces into tabs. If you wish to have a value with whitespace, the KEY:VALUE must be enclosed in double quotes. Refer to the start documen tation for complete input details. If not provided, value will be autogenerated based on the reads1 file basename." }
  SampleID: string
  std_dev: long?
  avg_frag_len: int?

outputs:
  abundance_out:
    type: File
    outputBinding:
      glob: '*.abundance.tsv.gz'
